pkg/segment/set.go

```diff
@@ -327,13 +332,27 @@ func (s *Set) Insert(gap GapIterator, r Range, val Value) Iterator {
 	}
 	if prev.Ok() && prev.End() == r.Start {
 		if mval, ok := (Functions{}).Merge(prev.Range(), prev.Value(), r, val); ok {
-			prev.SetEndUnchecked(r.End)
-			prev.SetValue(mval)
+			if prev.node.hasChildren {
+				// Update maxGap of this gap if the actual merge happens on a non-leaf node.
+				prev.SetEndUnchecked(r.End)
```

Comment:
This is actually the cleanest format because the SetEnd and SetValue will affect the if condition so it should be behind the if while updateMaxGapLeaf must be after these updates.

Reply:
Yes you are right. This part is polished.